### PR TITLE
added approximate routine for calculating position of the moon

### LINKED ISSUE #4943: Add approximate version of `get_moon`
Mostly to remind us of @StuartLittlefair's suggestion with an actual implementation: https://github.com/astropy/astropy/pull/4920#issuecomment-219773572

p.s. It might still be useful to check if `erfa` hacks give comparable precision.

### PR DESCRIPTION
**Pull request by StuartLittlefair**
This PR fixes #4943.

I have coded up the algorithm in Jean Meeus', 'Astronomical Algorithms', second edition, 1998, Willmann-Bell for the approximate position of the Moon. Compared with JPL ephemerides this is only accurate to 15 arcsecs and 60 km over the range 1950-2050 (see below). However, it does not require any public downloads. 

meeus

The function defined here has not been made part of the public API. I am thinking that the function can be used to add the moon to the ```approximate``` ephemeris added in #4920 once that PR is merged.

### CODE REVIEW

**Comment by StuartLittlefair** [comment]
@mhvk I guess now #4920 is in I should rebase this on astropy/master and make changes so the moon is available as part of the ```builtin``` ephemeris?

File: astropy/coordinates/meeus.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    F = coord.Angle(F).wrap_at(360*u.deg)
+
+    A1 = coord.Angle(u.deg*polyval(T, _coA1)).wrap_at(360*u.deg)
+    A2 = coord.Angle(u.deg*polyval(T, _coA2)).wrap_at(360*u.deg)
+    A3 = coord.Angle(u.deg*polyval(T, _coA3)).wrap_at(360*u.deg)
+    E = polyval(T, _coE)
+
+    suml = sumr = 0.0
+    lrnum = len(_MOON_L_R)
+    for i in range(lrnum):
```
Comment (mhvk) [inline astropy/coordinates/meeus.py]: A bit more pythonic might be:
```
for moon_l_r in _MOON_L_R:
    DNum, MNum, McNum, FNum, LFac, RFac = moon_l_r
    corr = E ** abs(MNum)
    suml += LFac * corr * np.sin(...)
    sumr += RFac * corr * np.cos(...)
```

File: astropy/coordinates/meeus.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    if location is None:
+        location = t.location
+
+    # number of centuries since J2000.0.
+    # This should strictly speaking be in Ephemeris Time, but TDB or TT
+    # will introduce error smaller than intrinsic accuracy of algorithm.
+    T = (t.tdb.jyear-2000.0)/100.
+
+    # constants that are needed for all calculations
+    Lc = u.deg*polyval(T, _coLc)
```
Comment (mhvk) [inline astropy/coordinates/meeus.py]: You can initialize as a `Quantity` while constructing the angle, so I would either remove the `u.deg*` here, or replace the two incantations with one (faster) one:
```
Lc = coord.Angle(polyval(T, _colc), u.deg, copy=False)
```
I'm not quite sure why the wrapping is necessary, but if it really is, you could also just initialize a `Longitude`, which wraps at 360 degrees by default.

File: astropy/coordinates/meeus.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    D = u.deg*polyval(T, _coD)
+    M = u.deg*polyval(T, _coM)
+    Mc = u.deg*polyval(T, _coMc)
+    F = u.deg*polyval(T, _coF)
+
+    # convert to angle objects and wrap at 360 degrees
```
Comment (mhvk) [inline astropy/coordinates/meeus.py]: As an alternative, maybe just remove the whole `Angle` conversion? Isn't a `Quantity` in degrees good enough?

**Comment by eteq** [comment]
@StuartLittlefair - yep, rebase just as you said (and maybe address @mhvk's comments at the same time?)

Two larger scale thoughts (I'll do a more thorough review post-rebase):
* I don't think there should be a `meeus.py`, because there's already some of those algorithms implemented elsewhere (either directly or indirectly via slightly different implementations of the same ideas).  What you've done here should instead be `orbital_elements.py` or similar.
* Along related lines: you've implemented almost everything needed to get a location with *arbitrary* orbital elements.  I think this is worth generalizing in that manner.  That could be done here (if you like), or as a follow-on PR (which I'm happy to do myself if you'd rather not, @StuartLittlefair).

**Comment by StuartLittlefair** [comment]
The commits I just pushed rebase this on @mhvk's changes made in #4920.

@eteq if these tests pass I guess this could do with a more thorough review. I thought about your suggestions for generalising to a location with arbitrary orbital elements but struggled to see how such a generalisation could be done. For example, the VSOP87 planetary theory uses different types of sum and even terms to the ELP2000-82 Lunar position model encoded here. If you think it can be done, however, feel free to make a PR and show me how!

File: astropy/coordinates/orbital_elements.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    Mc = u.Quantity(polyval(T, _coMc), u.deg)
+    F = u.Quantity(polyval(T, _coF), u.deg)
+
+    A1 = u.Quantity(polyval(T, _coA1), u.deg)
+    A2 = u.Quantity(polyval(T, _coA2), u.deg)
+    A3 = u.Quantity(polyval(T, _coA3), u.deg)
+    E = polyval(T, _coE)
+
+    suml = sumr = 0.0
+    for moon_l_r in _MOON_L_R:
```
Comment (mhvk) [inline astropy/coordinates/orbital_elements.py]: Looking at this again, I think one can simply write
```
for DNum, MNum, McNum, FNum, LFac, RFac in _MOON_L_R:
```
(and same below)

Sorry that I didn't note that earlier!

File: astropy/coordinates/orbital_elements.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+        corr = E ** abs(MNum)
+        sumb += BFac*corr*np.sin(D*DNum+M*MNum+Mc*McNum+F*FNum)
+
+    suml += (3958*np.sin(A1) + 1962*np.sin(Lc-F) + 318*np.sin(A2))
+    sumb += (-2235*np.sin(Lc) + 382*np.sin(A3) + 175*np.sin(A1-F) +
+             175*np.sin(A1+F) + 127*np.sin(Lc-Mc) - 115*np.sin(Lc+Mc))
+
+    # ensure units
+    suml = suml*u.deg
```
Comment (mhvk) [inline astropy/coordinates/orbital_elements.py]: The divisions below suggest the real units of `suml` and `sumb` are `u.microdegree`; I would use that (or move the division here).

File: astropy/coordinates/solar_system.py [hunk 6]
```diff
@@ -274,6 +276,13 @@ def _get_apparent_body_position(body, time, ephemeris):
     cartesian_position : `~astropy.coordinates.CartesianRepresentation`
         Barycentric (ICRS) apparent position of the body in cartesian coordinates
     """
+    # builtin ephemeris and moon is a special case, with no need to account for
```
Comment (mhvk) [inline astropy/coordinates/solar_system.py]: The comment should go two lines below. Also, maybe explicitly note that there is no need to account for light travel time *given* that the effect is smaller than the accuracy of the algorithm? (Or is it included already?)

File: astropy/coordinates/tests/test_solar_system.py [hunk 7]
```diff
@@ -207,7 +208,8 @@ def setup(self):
     def test_erfa_planet(self, body, sep_tol, dist_tol):
         """Test predictions using erfa/plan94.
 
-        Accuracies are maximum deviations listed in erfa/plan94.c
+        Accuracies are maximum deviations listed in erfa/plan94.c, for Jupiter and
+        Mercury, and that quoted in Meeus "Astronomical Algorithms" (1998) for the Moon.
         """
         # Add uncertainty in position of Earth
         dist_tol = dist_tol + 1300 * u.km
```
Comment (mhvk) [inline astropy/coordinates/tests/test_solar_system.py]: Is this needed for the Moon? I think I would just add it to `Mercury` in the list above as done for the other test (sorry, *I* should have done that!)

File: astropy/coordinates/tests/test_solar_system.py [hunk 8]
```diff
@@ -207,7 +208,8 @@ def setup(self):
     def test_erfa_planet(self, body, sep_tol, dist_tol):
```
Comment (mhvk) [inline astropy/coordinates/tests/test_solar_system.py]: Don't forget to actually add the moon in the line above!

File: docs/coordinates/solarsystem.rst [hunk 9]
```diff
@@ -92,6 +92,7 @@ For a list of the bodies for which positions can be calculated, do::
    'mercury',
    'venus',
    'earth-moon-barycenter',
+   'moon',
```
Comment (mhvk) [inline docs/coordinates/solarsystem.rst]: How about adding an example in the "one-off" section just above?

File: astropy/coordinates/tests/test_solar_system.py (hunk 8 above)
Comment (StuartLittlefair) [inline astropy/coordinates/tests/test_solar_system.py]: Oops - good spot @mhvk. This is in the wrong place! I didn't intend to add Kitt Peak tests for the Moon, as these depend on #4986 being merged to pass...

**Comment by eteq** [comment]
I still think we might want to generalize the orbital elements machinery, *but* on reflection it's definitely a task for a future PR. So with that this looks good-to-go to me, right @StuartLittlefair?

 @astrofrog and @taldcroft, are you OK with backporting this to 1.2 (as it's part of the solar system stuff that's new in 1.2).  If so, it doesn't need a changelog entry as it's an addition to an already-mentioned new feature.

**Comment by astrofrog** [comment]
:+1: to backporting

**Comment by astrofrog** [comment]
@StuartLittlefair - can you confirm this is ready to go?

@taldcroft - do you agree with backporting this?

**Comment by StuartLittlefair** [comment]
@mhvk - yes, ready to go from me.

**Comment by taldcroft** [comment]
👍 on backporting to 1.2.

**Pull request merged**